### ISSUE TITLE
Dockerode features don't work with TCP daemon + certificates

### ISSUE THREAD

**Issue by atomi** [microsoft/vscode-docker#2058]
I've setup a docker context using 

``` bash
docker context create machine --docker "host=tcp://192.168.99.120:2376,ca=/Machines/machines/default/ca.pem,cert=/Machines/machines/default/cert.pem,key=/Machines/machines/default/key.pem" --default-stack-orchestrator swarm
```
This works in my OS terminal and VSCode terminal, however VSCode Docker plugin says 
"Failed to connect. Is Docker running?"

I can have VSCode Docker plugin connect to my machine if I instead use this in my settings
```
  "docker.machineName": "machine",
  "docker.certPath": "Machines/machines/default",
  "docker.host": "tcp://192.168.99.120:2376",
```

But again, not with Docker context.

**Comment by bwateratmsft** [comment]
@atomi because of what a mess the CA setup is we did not implement it for Dockerode support. You can use the `docker.dockerodeOptions` setting to set whatever you want to be passed to Dockerode, including CA settings.

**Comment by thofil73** [comment]
Is there any progress on this issue?
I am dealing with several docker contexts, some use SSH, others use TLS.
SSH is working fine, but TLS fails to connect as mentioned by atomi.

The two workarounds from above don't work for me, as they would make all my contexts disappear except for the 1 and only "hard-coded" connection.

IMHO, instead of VS Code Docker plugin using the explicit settings
     "docker.certPath": "Machines/machines/default",
     "docker.host": "tcp://192.168.99.120:2376",
it could
     1) run 'docker context inspect <current context>'
     2) check whether "TLSMaterial" : "docker": [...] is non-empty => TLS
     3) in case of TLS,
           a) set docker.host := value from "Endpoints" : "docker" : "Host"
           b) set docker.certPath := value from "Storage" : "TLSPath"

which would allow me to smoothly switch between my docker contexts.

Is my thinking too simple?

**Comment by bwateratmsft** [comment]
@thofil73 we're probably going to proceed with #3263, which as a side effect would fix this issue--basically, anything that the CLI could reach, would be reachable from the extension. Dockerode would no longer be needed.

**Comment by LouisStAmour-TELUS** [comment]
Just a post that I tried setting up a Minikube context with:

    docker context create minikube \
        --description "Minikube" \
        --default-stack-orchestrator kubernetes \
        --docker "host=tcp://127.0.0.1:64348,ca=$HOME/.minikube/certs/ca.pem,cert=$HOME/.minikube/certs/cert.pem,key=$HOME/.minikube/certs/key.pem,skip-tls-verify=false" \
        --kubernetes "config-file=$HOME/.kube/config"

And also saw `Failed to connect. Is Docker running? Error: (HTTP code 400) bad parameter - Client sent an HTTP request to an HTTPS server.`

The only downside to using native Docker CLI instead is that I couldn't seem to configure my Docker CLI to connect via the context either. I tried `docker context use minikube` but all I got was an asterisk next to the default config indicating it was the one in use when I ran `docker context list`. Docker contexts are incredibly frustrating.

Perhaps we can invent a way to switch docker configs within the Docker extension that doesn't use contexts? I'd like to be able to switch between Docker for Mac and Minikube within the extension but it doesn't seem easy/possible with the default configs. Minikube itself only supports environment variables via the `minikube docker-env` command's output. This is starting to feel like a different ask, that #2681 might be more appropriate to handle, namely to support multiple extension settings for docker endpoints and switch between them as an alternative to using contexts?

**Comment by LouisStAmour** [comment]
Just a post that I tried setting up a Minikube context with:

    docker context create minikube \
        --description "Minikube" \
        --default-stack-orchestrator kubernetes \
        --docker "host=tcp://127.0.0.1:64348,ca=$HOME/.minikube/certs/ca.pem,cert=$HOME/.minikube/certs/cert.pem,key=$HOME/.minikube/certs/key.pem,skip-tls-verify=false" \
        --kubernetes "config-file=$HOME/.kube/config"

And also saw `Failed to connect. Is Docker running? Error: (HTTP code 400) bad parameter - Client sent an HTTP request to an HTTPS server.`

The only downside to using native Docker CLI instead is that I couldn't seem to configure my Docker CLI to connect via the context either. I tried `docker context use minikube` but all I got was an asterisk next to the default config indicating it was the one in use when I ran `docker context list`. Docker contexts are incredibly frustrating.

Perhaps we can invent a way to switch docker configs within the Docker extension that doesn't use contexts? I'd like to be able to switch between Docker for Mac and Minikube within the extension but it doesn't seem easy/possible with the default configs. Minikube itself only supports environment variables via the `minikube docker-env` command's output. This is starting to feel like a different ask, that #2681 might be more appropriate to handle, namely to support multiple extension settings for docker endpoints and switch between them as an alternative to using contexts?
